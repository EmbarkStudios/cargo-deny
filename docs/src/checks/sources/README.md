# sources

The sources check ensures crates only come from sources you trust.

```bash
cargo deny check sources
```

<img src="https://imgur.com/xdHFDWS.png"/>

## Use Case - Only allowing known/trusted sources

Cargo can retrieve crates from a variety of sources, namely registries, git 
repositories, or local file paths. This is great in general and very flexible 
for development. But, especially when re-routing dependencies to git 
repositories, increasing the amount of sources that a project has to trust may
be something a repository wants to explicitly opt-in to.

See [Why npm lockfiles can be a security blindspot for injecting malicious 
modules](https://snyk.io/blog/why-npm-lockfiles-can-be-a-security-blindspot-for-injecting-malicious-modules/)
for the motivating reason for why this check was added.

## Use Case - Only using vendored file dependencies

A project may want to only support local file dependencies, such as having all 
dependencies vendored into the repository for full control and offline building.
This can be achieved by disallowing all git and registry sources to ensure that
every dependency is added into your source control rather than via an external
source.
