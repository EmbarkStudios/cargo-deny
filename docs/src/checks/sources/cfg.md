# The `[sources]` section

Contains all of the configuration for `cargo deny check sources`

## Example Config

```ini
{{#include ../../../../tests/cfg/sources.toml}}
```

### The `unknown-registry` field (optional)

Determines what happens when a crate from a crate registry that is not in the `allow-registry` list is encountered.

* `deny` - Will emit an error with the URL of the source, and fail the check.
* `warn` (default) - Prints a warning for each crate, but does not fail the check.
* `allow` - Prints a note for each crate, but does not fail the check.

### The `unknown-git` field (optional)

Determines what happens when a crate from a git repository not in the `allow-git` list is encountered.

* `deny` - Will emit an error with the URL of the repository, and fail the check.
* `warn` (default) - Prints a warning for each crate, but does not fail the check.
* `allow` - Prints a note for each crate, but does not fail the check.

### The `required-git-spec` (optional)

Determines which [specifiers](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-git-repositories) are required for git sources. Git sources are a convient way to use patched code temporarily, but they have downsides for long term maintenance, as the specifier you use for the source determines what happens when you do a `cargo update`, and in the default case, this means you essentially have a wildcard dependency on the repository.

This configuration value allows you to control what specifiers you want to allow for your git sources to reduce surprises. The following values are listed in order from least to most specific, and using a less specific specifier will also allow all of the more specific ones.

* `any` (default) - Allows all git specs, including the default of not having
any specifier, which tracks the latest commit on the `master` branch of the repo
* `branch` - Allows the `branch = "<branch_name>"` specifier.
* `tag` - Allows the `tag = "<tag_name>"` specifier.
* `rev` - Allows the `rev = "<commit_sha>"` specifier.

### The `allow-git` field (optional)

Configure which git urls are allowed for crate sources. If a crate's source is not in one of the listed urls, then the `unknown-git` setting will determine how it is handled.

### The `allow-registry` field (optional)

The list of registries that are allowed. If a crate is not found in one of the listed registries, then the `unknown-registry` setting will determine how it is handled.

If not specified, this list will by default contain the [crates.io](http://crates.io) registry, equivalent to this:

```ini
[sources]
allow-registry = [
    "https://github.com/rust-lang/crates.io-index"
]
```

To not allow any crates registries, set it to empty:

```ini
[sources]
unknown-registry = "deny"
allow-registry = []
```

### The `allow-org` field (optional)

Generally, I think most projects in the Rust space probably follow a similar procedure as we do when they want to fix a bug or add a feature to one of their dependencies, which is basically.

1. Fork the crate to make your changes
1. Hack away locally, probably just patching your project(s) to use a `path` dependency to the cloned fork
1. Push changes to your fork, and once you're happy, change the `path` dependency to a `git` dependency and point it to your fork for others/CI to be able to use the same changes easily
1. Eventually (hopefully!) make a PR to the original repo with your changes
1. Hopefully get your changes merged to the original repo
1. Wait until a release is made that incorporates your changes, possibly changing the `git` source to point to the original repo
1. Remove the `git` source and instead point at the new version of the crate with your changes
1. Profit!

When working in a company or organization, it is often the case that all crates will be forked to a shared organization account rather than a personal Github account. However, if you lint your git sources, every new and deleted fork needs to keep that list updated, which is tedious, even if all the forks fall under the same organization (in Github terminology), even though presumably only people you trust have permission to create forks there, and you would like to just blanket trust any repo under that org.

The `allow-org` object allows you to specify 1 or more orgs in several VCS providers to more easily configure git sources for your projects.

#### The `github` field (optional)

Allows you to specify one or more `github.com` organizations to allow as git sources.

```ini
[sources.allow-org]
github = ["YourCoolOrgGoesHere"]
```

#### The `gitlab` field (optional)

Allows you to specify one or more `gitlab.com` organizations to allow as git sources.

```ini
[sources.allow-org]
gitlab = ["YourCoolOrgGoesHere"]
```

#### The `bitbucket` field (optional)

Allows you to specify one or more `bitbucket.org` organizations to allow as git sources.

```ini
[sources.allow-org]
bitbucket = ["YourCoolOrgGoesHere"]
```
