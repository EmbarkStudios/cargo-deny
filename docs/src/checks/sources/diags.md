# Sources diagnostics

### `git-source-underspecified`

A `git` source uses a specification that doesn't meet the minimum specifier required by [`sources.required-git-spec`](cfg.md#the-required-git-spec-optional).

### `allowed-source`

A crate source is explicitly allowed by [`sources.allow-git`](cfg.md#the-allow-git-field-optional) or [`sources.allow-registry`](cfg.md#the-allow-registry-field-optional).

### `allowed-by-organization`

A crate source was explicitly allowed by an entry in [`sources.allow-org`](cfg.md#the-allow-org-field-optional).

### `source-not-allowed`

A crate's source was not explicitly allowed.

### `unmatched-source`

An allowed source in [`sources.allow-git`](cfg.md#the-allow-git-field-optional) or [`sources.allow-registry`](cfg.md#the-allow-registry-field-optional) was not encountered.

### `unmatched-organization`

An allowed source in [`sources.allow-org`](cfg.md#the-allow-org-field-optional) was not encountered.
