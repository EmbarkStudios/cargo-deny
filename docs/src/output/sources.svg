<svg fill="none" viewBox="0 0 850 428" width="850" height="428" xmlns="http://www.w3.org/2000/svg">
  <foreignObject width="100%" height="100%">
    <div xmlns="http://www.w3.org/1999/xhtml">

      <style>.term-container {
  background: #282c34;
  border-radius: 5px;
  color: #abb2bf;
  word-break: break-word;
  overflow-wrap: break-word;
  font-family: Monaco, courier, monospace;
  font-size: 12px;
  line-height: 20px;
  padding: 14px 18px;
  white-space: pre-wrap;
}

.term-fg1 { font-weight: bold; }
.term-fg3 { font-style: italic; } /* italic */
.term-fg4 { text-decoration: underline; } /* underline */
.term-fg9 { text-decoration: line-through; } /* crossed-out */

.term-fg30 { color: #666; }
.term-fg31 { color: #e06c75; } /* red */
.term-fg32 { color: #98c379; } /* green */
.term-fg33 { color: #e5c07b; } /* yellow */
.term-fg34 { color: #61afef; } /* blue */
.term-fg35 { color: #c678dd; } /* magenta */
.term-fg36 { color: #56b6c2; } /* cyan */
.term-fgx9 { color: #be5046; } /* bright red */

      </style>

      <div class="term-container">cargo-deny --manifest-path examples&#47;07_deny_sources&#47;Cargo.toml --color always check sources`
<span class="term-fgx9 term-fg1">error[source-not-allowed]</span><span class="term-fg1">: detected &#39;git&#39; source not explicitly allowed</span>
  <span class="term-fg34">┌─</span> &#47;home&#47;jake&#47;code&#47;cargo-deny&#47;examples&#47;07_deny_sources&#47;Cargo.lock:7:16
  <span class="term-fg34">│</span>
<span class="term-fg34">7</span> <span class="term-fg34">│</span> bitflags 1.2.1 <span class="term-fg31">git+https:&#47;&#47;github.com&#47;bitflags&#47;bitflags.git#1e41c2959196b374812450037f55c492d3911465</span>
  <span class="term-fg34">│</span>                <span class="term-fg31">^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^</span> <span class="term-fg31">source</span>
  <span class="term-fg34">│</span>
  <span class="term-fg34">=</span> bitflags v1.2.1
    └── sources v0.1.0

<span class="term-fgx9 term-fg1">error[source-not-allowed]</span><span class="term-fg1">: detected &#39;git&#39; source not explicitly allowed</span>
   <span class="term-fg34">┌─</span> &#47;home&#47;jake&#47;code&#47;cargo-deny&#47;examples&#47;07_deny_sources&#47;Cargo.lock:83:18
   <span class="term-fg34">│</span>
<span class="term-fg34">83</span> <span class="term-fg34">│</span> mlua_serde 0.6.0 <span class="term-fg31">git+https:&#47;&#47;tulpa.dev&#47;lua&#47;mlua_serde#28ab03a516dae6e68062e6613442582f6718ae96</span>
   <span class="term-fg34">│</span>                  <span class="term-fg31">^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^</span> <span class="term-fg31">source</span>
   <span class="term-fg34">│</span>
   <span class="term-fg34">=</span> mlua_serde v0.6.0
     └── sources v0.1.0

sources <span class="term-fg31">FAILED</span></div>

    </div>
  </foreignObject>
</svg>
