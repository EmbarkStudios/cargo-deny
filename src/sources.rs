mod cfg;
mod diags;
pub use cfg::{Config, GitSpec, ValidConfig};
pub use diags::Code;

use crate::{
    diag::{CfgCoord, Check, ErrorSink, Label, Pack},
    LintLevel,
};

const CRATES_IO_URL: &str = "https://github.com/rust-lang/crates.io-index";

pub fn check(ctx: crate::CheckCtx<'_, ValidConfig>, sink: impl Into<ErrorSink>) {
    use bitvec::prelude::*;

    // early out if everything is allowed
    if ctx.cfg.unknown_registry == LintLevel::Allow && ctx.cfg.unknown_git == LintLevel::Allow {
        return;
    }

    let mut sink = sink.into();

    // scan through each crate and check the source of it

    // keep track of which sources are actually encountered, so we can emit a
    // warning if the user has listed a source that no crates are actually using
    let mut source_hits: BitVec = BitVec::repeat(false, ctx.cfg.allowed_sources.len());
    let mut org_hits: BitVec = BitVec::repeat(false, ctx.cfg.allowed_orgs.len());

    let min_git_spec = ctx.cfg.required_git_spec.as_ref().map(|rgs| {
        (
            rgs.value,
            CfgCoord {
                span: rgs.span.clone(),
                file: ctx.cfg.file_id,
            },
        )
    });

    for (i, krate) in ctx.krates.krates().enumerate() {
        let source = match &krate.source {
            Some(source) => source,
            None => continue,
        };

        let mut pack = Pack::with_kid(Check::Sources, krate.id.clone());

        let source_label = {
            let mut span = ctx.krate_spans[i].clone();

            // The krate span is the complete id, but we only want
            // to highlight the source component
            let last_space = krate.id.repr.rfind(' ').unwrap();

            span.start = span.start + last_space + 1;
            Label::primary(ctx.krate_spans.file_id, span).with_message("source")
        };

        // get allowed list of sources to check
        let (lint_level, type_name) = if source.is_registry() {
            (ctx.cfg.unknown_registry, "registry")
        } else if let Some(spec) = source.git_spec() {
            // Ensure the git source has at least the minimum specification
            if let Some((min, cfg_coord)) = &min_git_spec {
                if spec < *min {
                    pack.push(diags::BelowMinimumRequiredSpec {
                        src_label: &source_label,
                        min_spec: *min,
                        actual_spec: spec,
                        min_spec_cfg: cfg_coord.clone(),
                    });
                }
            }

            (ctx.cfg.unknown_git, "git")
        } else {
            continue;
        };

        // check if the source URL is in the list of allowed sources
        let diag: crate::diag::Diag = if let Some(ind) = ctx
            .cfg
            .allowed_sources
            .iter()
            .position(|src| krate.matches_url(&src.url.value, src.exact))
        {
            source_hits.as_mut_bitslice().set(ind, true);

            // Show the location of the config that allowed this source, unless
            // it's crates.io since that will be a vast majority of crates and
            // is the default, so we might not have a real source location anyways
            if krate.is_crates_io() {
                continue;
            }

            diags::ExplicitlyAllowedSource {
                src_label: &source_label,
                type_name,
                allow_cfg: CfgCoord {
                    file: ctx.cfg.file_id,
                    span: ctx.cfg.allowed_sources[ind].url.span.clone(),
                },
            }
            .into()
        } else if let Some((orgt, orgname)) = krate.source.as_ref().and_then(|s| {
            let crate::Source::Git { url, .. } = s else { return None; };
            get_org(url)
        }) {
            if let Some(ind) = ctx
                .cfg
                .allowed_orgs
                .iter()
                .position(|(sorgt, sorgn)| orgt == *sorgt && sorgn.value.as_str() == orgname)
            {
                org_hits.as_mut_bitslice().set(ind, true);
                diags::SourceAllowedByOrg {
                    src_label: &source_label,
                    org_cfg: CfgCoord {
                        file: ctx.cfg.file_id,
                        span: ctx.cfg.allowed_orgs[ind].1.span.clone(),
                    },
                }
                .into()
            } else {
                diags::SourceNotExplicitlyAllowed {
                    src_label: &source_label,
                    lint_level,
                    type_name,
                }
                .into()
            }
        } else {
            diags::SourceNotExplicitlyAllowed {
                src_label: &source_label,
                lint_level,
                type_name,
            }
            .into()
        };

        pack.push(diag);
        sink.push(pack);
    }

    let mut pack = Pack::new(Check::Sources);

    for src in source_hits
        .into_iter()
        .zip(ctx.cfg.allowed_sources.into_iter())
        .filter_map(|(hit, src)| if !hit { Some(src) } else { None })
    {
        // If someone in is in a situation that they want to disallow crates
        // from crates.io, they should set the allowed registries manually
        if src.url.as_ref().as_str() == CRATES_IO_URL {
            continue;
        }

        pack.push(diags::UnmatchedAllowSource {
            allow_src_cfg: CfgCoord {
                span: src.url.span,
                file: ctx.cfg.file_id,
            },
        });
    }

    for (org_type, orgs) in org_hits
        .into_iter()
        .zip(ctx.cfg.allowed_orgs.into_iter())
        .filter_map(|(hit, src)| if !hit { Some(src) } else { None })
    {
        pack.push(diags::UnmatchedAllowOrg {
            allow_org_cfg: CfgCoord {
                span: orgs.span,
                file: ctx.cfg.file_id,
            },
            org_type,
        });
    }

    if !pack.is_empty() {
        sink.push(pack);
    }
}

#[derive(PartialEq, Eq, Debug, Copy, Clone)]
pub enum OrgType {
    Github,
    Gitlab,
    Bitbucket,
}

use std::fmt;
impl fmt::Display for OrgType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(match self {
            Self::Github => "github.com",
            Self::Gitlab => "gitlab.com",
            Self::Bitbucket => "bitbucket.org",
        })
    }
}

fn get_org(url: &url::Url) -> Option<(OrgType, &str)> {
    url.domain().and_then(|domain| {
        let org_type = if domain.eq_ignore_ascii_case("github.com") {
            OrgType::Github
        } else if domain.eq_ignore_ascii_case("gitlab.com") {
            OrgType::Gitlab
        } else if domain.eq_ignore_ascii_case("bitbucket.org") {
            OrgType::Bitbucket
        } else {
            return None;
        };

        url.path_segments()
            .and_then(|mut f| f.next())
            .map(|org| (org_type, org))
    })
}
