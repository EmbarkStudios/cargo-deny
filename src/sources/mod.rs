mod cfg;
pub use cfg::{Config, ValidConfig};

use crate::{
    diag::{Diagnostic, Label, Pack, Severity},
    LintLevel,
};

pub fn check(ctx: crate::CheckCtx<'_, ValidConfig>, sender: crossbeam::channel::Sender<Pack>) {
    use bitvec::prelude::*;

    // early out if everything is allowed
    if ctx.cfg.unknown_registry == LintLevel::Allow && ctx.cfg.unknown_git == LintLevel::Allow {
        return;
    }

    // scan through each crate and check the source of it

    // keep track of which sources are actually encountered, so we can emit a
    // warning if the user has listed a source that no crates are actually using
    let mut source_hits = bitvec![0; ctx.cfg.allowed_sources.len()];

    for (i, krate) in ctx.krates.krates().map(|kn| &kn.krate).enumerate() {
        let source = match &krate.source {
            Some(source) => source,
            None => continue,
        };

        // get URL without git revision (query & fragment)
        // example URL in Cargo.lock: https://github.com/RustSec/rustsec-crate.git?rev=aaba369#aaba369bebc4fcfb9133b1379bcf430b707188a2
        // where we only want:        https://github.com/RustSec/rustsec-crate.git
        let source_url = {
            let mut url = source.url().clone();
            url.set_query(None);
            url.set_fragment(None);
            url
        };

        // get allowed list of sources to check

        let (lint_level, type_name) = if source.is_registry() {
            (ctx.cfg.unknown_registry, "registry")
        } else if source.is_git() {
            (ctx.cfg.unknown_git, "git")
        } else {
            continue;
        };

        // check if the source URL is list of allowed sources
        match ctx
            .cfg
            .allowed_sources
            .iter()
            .position(|src| src == &source_url)
        {
            Some(ind) => source_hits.as_mut_bitslice().set(ind, true),
            None => {
                let mut span = ctx.krate_spans[i].clone();

                // The krate span is the complete id, but we only want
                // to highlight the source component
                let last_space = krate.id.repr.rfind(' ').unwrap();

                span.start = span.start + last_space as u32 + 1;

                let mut pack = Pack::with_kid(krate.id.clone());
                pack.push(Diagnostic::new(
                    match lint_level {
                        LintLevel::Warn => Severity::Warning,
                        LintLevel::Deny => Severity::Error,
                        LintLevel::Allow => Severity::Note,
                    },
                    format!("detected '{}' source not specifically allowed", type_name,),
                    Label::new(ctx.spans_id, span, "source"),
                ));

                sender.send(pack).unwrap();
            }
        }
    }

    for src in source_hits
        .into_iter()
        .zip(ctx.cfg.allowed_sources.into_iter())
        .filter_map(|(hit, src)| if !hit { Some(src) } else { None })
    {
        sender
            .send(
                Diagnostic::new(
                    Severity::Warning,
                    "allowed source was not encountered",
                    Label::new(
                        ctx.cfg.file_id,
                        src.span,
                        "no crate source matched these criteria",
                    ),
                )
                .into(),
            )
            .unwrap();
    }
}
