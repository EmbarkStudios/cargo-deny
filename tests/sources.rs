use cargo_deny::sources;

#[macro_use]
mod utils;

#[test]
fn fails_unknown_git() {
    let cfg = "unknown-git = 'deny'";

    let krates = utils::get_test_data_krates("sources").unwrap();
    let diags = utils::gather_diagnostics::<sources::Config, _, _>(
        krates,
        "fails_unknown_git",
        Some(cfg),
        None,
        |ctx, _, tx| {
            sources::check(ctx, tx);
        },
    )
    .unwrap();

    let failed_urls = [
        // Note this one is used by multiple crates, but that's ok
        "https://gitlab.com/amethyst-engine/amethyst",
        "https://github.com/EmbarkStudios/krates",
        "https://bitbucket.org/marshallpierce/line-wrap-rs",
        "https://github.com/EmbarkStudios/spdx",
    ];

    for url in &failed_urls {
        assert!(
            diags.iter().any(|v| {
                field_eq!(v, "/fields/severity", "error")
                    && field_eq!(
                        v,
                        "/fields/message",
                        "detected 'git' source not specifically allowed"
                    )
                    && v.pointer("/fields/labels/0/span")
                        .unwrap()
                        .as_str()
                        .unwrap()
                        .contains(url)
            }),
            "failed to find diagnostic for git source '{}'",
            url
        );
    }
}

#[test]
fn allows_git() {
    let cfg = "unknown-git = 'deny'
    allow-git = [
        'https://gitlab.com/amethyst-engine/amethyst',
        'https://github.com/EmbarkStudios/krates',
        'https://bitbucket.org/marshallpierce/line-wrap-rs',
    ]";

    let krates = utils::get_test_data_krates("sources").unwrap();
    let diags = utils::gather_diagnostics::<sources::Config, _, _>(
        krates,
        "fails_unknown_git",
        Some(cfg),
        None,
        |ctx, _, tx| {
            sources::check(ctx, tx);
        },
    )
    .unwrap();

    let allowed_urls = [
        // Note this one is used by multiple crates, but that's ok
        "https://gitlab.com/amethyst-engine/amethyst",
        "https://github.com/EmbarkStudios/krates",
        "https://bitbucket.org/marshallpierce/line-wrap-rs",
    ];

    for url in &allowed_urls {
        assert!(
            diags.iter().any(|v| {
                field_eq!(v, "/fields/severity", "note")
                    && field_eq!(v, "/fields/message", "\'git\' source explicitly allowed")
                    && v.pointer("/fields/labels/0/span")
                        .unwrap()
                        .as_str()
                        .unwrap()
                        .contains(url)
                    && field_eq!(v, "/fields/labels/1/span", format!("'{}'", url))
            }),
            "failed to find diagnostic for git source '{}'",
            url
        );
    }
}

#[test]
fn allows_github_org() {
    // We shouldn't have any errors for the embark urls now
    let cfg = "unknown-git = 'deny'
    [allow-org]
    github = ['EmbarkStudios']
    ";

    let krates = utils::get_test_data_krates("sources").unwrap();
    let diags = utils::gather_diagnostics::<sources::Config, _, _>(
        krates,
        "allows_github_org",
        Some(cfg),
        None,
        |ctx, _, tx| {
            sources::check(ctx, tx);
        },
    )
    .unwrap();

    let allowed_by_org = [
        "https://github.com/EmbarkStudios/krates",
        "https://github.com/EmbarkStudios/spdx",
    ];

    for diag in diags {
        match diag.pointer("/fields/severity").unwrap().as_str().unwrap() {
            "error" => {
                let source = diag
                    .pointer("/fields/labels/0/span")
                    .unwrap()
                    .as_str()
                    .unwrap();

                assert!(!allowed_by_org.iter().any(|ao| source.contains(ao)));
            }
            "note" => {
                let source = diag
                    .pointer("/fields/labels/0/span")
                    .unwrap()
                    .as_str()
                    .unwrap();

                assert!(allowed_by_org.iter().any(|ao| source.contains(ao)));
            }
            ty => unreachable!("unexpected '{}' diagnostic", ty),
        }
    }
}

#[test]
fn allows_gitlab_org() {
    let cfg = "unknown-git = 'deny'
    [allow-org]
    gitlab = ['amethyst-engine']
    ";

    let krates = utils::get_test_data_krates("sources").unwrap();
    let diags = utils::gather_diagnostics::<sources::Config, _, _>(
        krates,
        "allows_gitlab_org",
        Some(cfg),
        None,
        |ctx, _, tx| {
            sources::check(ctx, tx);
        },
    )
    .unwrap();

    let allowed_by_org = ["https://gitlab.com/amethyst-engine"];

    for diag in diags {
        match diag.pointer("/fields/severity").unwrap().as_str().unwrap() {
            "error" => {
                let source = diag
                    .pointer("/fields/labels/0/span")
                    .unwrap()
                    .as_str()
                    .unwrap();

                assert!(!allowed_by_org.iter().any(|ao| source.contains(ao)));
            }
            "note" => {
                let source = diag
                    .pointer("/fields/labels/0/span")
                    .unwrap()
                    .as_str()
                    .unwrap();

                assert!(allowed_by_org.iter().any(|ao| source.contains(ao)));
            }
            ty => unreachable!("unexpected '{}' diagnostic", ty),
        }
    }
}

#[test]
fn allows_bitbucket_org() {
    let cfg = "unknown-git = 'deny'
    [allow-org]
    bitbucket = ['marshallpierce']
    ";

    let krates = utils::get_test_data_krates("sources").unwrap();
    let diags = utils::gather_diagnostics::<sources::Config, _, _>(
        krates,
        "allows_bitbucket_org",
        Some(cfg),
        None,
        |ctx, _, tx| {
            sources::check(ctx, tx);
        },
    )
    .unwrap();

    let allowed_by_org = ["https://bitbucket.org/marshallpierce/line-wrap-rs"];

    for diag in diags {
        match diag.pointer("/fields/severity").unwrap().as_str().unwrap() {
            "error" => {
                let source = diag
                    .pointer("/fields/labels/0/span")
                    .unwrap()
                    .as_str()
                    .unwrap();

                assert!(!allowed_by_org.iter().any(|ao| source.contains(ao)));
            }
            "note" => {
                let source = diag
                    .pointer("/fields/labels/0/span")
                    .unwrap()
                    .as_str()
                    .unwrap();

                assert!(allowed_by_org.iter().any(|ao| source.contains(ao)));
            }
            ty => unreachable!("unexpected '{}' diagnostic", ty),
        }
    }
}

#[test]
fn validates_git_source_specs() {
    use sources::GitSpec;

    assert!(GitSpec::Rev > GitSpec::Tag);
    assert!(GitSpec::Tag > GitSpec::Branch);
    assert!(GitSpec::Branch > GitSpec::Any);

    let levels = [
        (GitSpec::Rev, "https://gitlab.com/amethyst-engine/amethyst"),
        (GitSpec::Tag, "https://github.com/EmbarkStudios/spdx"),
        (GitSpec::Branch, "https://github.com/EmbarkStudios/krates"),
        (
            GitSpec::Any,
            "https://bitbucket.org/marshallpierce/line-wrap-rs",
        ),
    ];

    for (i, (spec, _url)) in levels.iter().enumerate() {
        let cfg = format!(
            "unknown-git = 'allow'
        required-git-spec = '{}'",
            spec
        );

        let krates = utils::get_test_data_krates("sources").unwrap();
        let mut diags = utils::gather_diagnostics::<sources::Config, _, _>(
            krates,
            "validates_git_source_specs",
            Some(&cfg),
            None,
            |ctx, _, tx| {
                sources::check(ctx, tx);
            },
        )
        .unwrap();

        diags.retain(|d| {
            d.pointer("/fields/message")
                .unwrap()
                .as_str()
                .unwrap()
                .starts_with("'git' source is underspecified, expected")
        });

        for (j, (_, url)) in levels.iter().enumerate() {
            let severities: Vec<_> = diags
                .iter()
                .filter_map(|d| {
                    d.pointer("/fields/labels/0/span")
                        .and_then(|u| u.as_str())
                        .and_then(|u| {
                            if u.contains(url) {
                                d.pointer("/fields/severity").and_then(|s| s.as_str())
                            } else {
                                None
                            }
                        })
                })
                .collect();

            if j <= i {
                assert!(severities.is_empty());
            } else {
                assert!(!severities.is_empty());
                assert!(severities.into_iter().all(|s| s == "error"));
            }
        }
    }
}
