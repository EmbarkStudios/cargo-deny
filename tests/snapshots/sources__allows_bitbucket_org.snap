---
source: tests/sources.rs
expression: diags
---
[
  {
    "fields": {
      "code": "source-not-allowed",
      "graphs": [
        {
          "Krate": {
            "name": "amethyst_core",
            "version": "0.10.1"
          },
          "parents": [
            {
              "Krate": {
                "kind": "build",
                "name": "sources",
                "version": "0.1.0"
              }
            }
          ]
        }
      ],
      "labels": [
        {
          "column": 22,
          "line": 7,
          "message": "source",
          "span": "git+https://gitlab.com/amethyst-engine/amethyst?rev=0c2da617#0c2da61772b89323af9dcfed0ed00b2a698d95b5"
        }
      ],
      "message": "detected 'git' source not explicitly allowed",
      "severity": "error"
    },
    "type": "diagnostic"
  },
  {
    "fields": {
      "code": "source-not-allowed",
      "graphs": [
        {
          "Krate": {
            "name": "amethyst_error",
            "version": "0.5.1"
          },
          "parents": [
            {
              "Krate": {
                "name": "amethyst_core",
                "version": "0.10.1"
              },
              "parents": [
                {
                  "Krate": {
                    "kind": "build",
                    "name": "sources",
                    "version": "0.1.0"
                  }
                }
              ]
            }
          ]
        }
      ],
      "labels": [
        {
          "column": 22,
          "line": 8,
          "message": "source",
          "span": "git+https://gitlab.com/amethyst-engine/amethyst?rev=0c2da617#0c2da61772b89323af9dcfed0ed00b2a698d95b5"
        }
      ],
      "message": "detected 'git' source not explicitly allowed",
      "severity": "error"
    },
    "type": "diagnostic"
  },
  {
    "fields": {
      "code": "source-not-allowed",
      "graphs": [
        {
          "Krate": {
            "name": "krates",
            "version": "0.4.1"
          },
          "parents": [
            {
              "Krate": {
                "name": "sources",
                "version": "0.1.0"
              }
            }
          ]
        }
      ],
      "labels": [
        {
          "column": 14,
          "line": 38,
          "message": "source",
          "span": "git+https://github.com/EmbarkStudios/krates?branch=main#3c1eab0f2322bbb16d1d26e82953f702b965e2e5"
        }
      ],
      "message": "detected 'git' source not explicitly allowed",
      "severity": "error"
    },
    "type": "diagnostic"
  },
  {
    "fields": {
      "code": "allowed-by-organization",
      "graphs": [
        {
          "Krate": {
            "name": "line-wrap",
            "version": "0.1.1"
          },
          "parents": [
            {
              "Krate": {
                "kind": "build",
                "name": "sources",
                "version": "0.1.0"
              }
            }
          ]
        }
      ],
      "labels": [
        {
          "column": 17,
          "line": 42,
          "message": "source",
          "span": "git+https://bitbucket.org/marshallpierce/line-wrap-rs#41e42b9d95684d41615dd3d0aeeb1cfc4cc47225"
        },
        {
          "column": 18,
          "line": 3,
          "message": "organization allowance",
          "span": "'marshallpierce'"
        }
      ],
      "message": "source allowed by organization allowance",
      "severity": "note"
    },
    "type": "diagnostic"
  },
  {
    "fields": {
      "code": "source-not-allowed",
      "graphs": [
        {
          "Krate": {
            "name": "spdx",
            "version": "0.3.4"
          },
          "parents": [
            {
              "Krate": {
                "kind": "dev",
                "name": "sources",
                "version": "0.1.0"
              }
            }
          ]
        }
      ],
      "labels": [
        {
          "column": 12,
          "line": 88,
          "message": "source",
          "span": "git+https://github.com/EmbarkStudios/spdx?tag=0.3.4#ba8dc9f0b164619771892b865f00236ebbd4cd39"
        }
      ],
      "message": "detected 'git' source not explicitly allowed",
      "severity": "error"
    },
    "type": "diagnostic"
  }
]
